### PR TITLE
fix: Fix PayloadProcessor response payload race condition

### LINKED ISSUE #111: AsyncPayloadProcessor owned payloads are released too early
## Issue:
AsyncPayloadProcessor sometimes sends null response payloads, despite the original infer caller getting a valid output back to the calling process.

## Triggers
If the payloads are processed instantly by the AsyncProcessor, the majority (but not all) responses arrive intact. If the AsyncProcessor is slow, the majority of responses arrive with 'null' data.

## Investigation
It looks like the response bytebufs are read too early somwehere, causing their `readerIndex` to equal their capacity at processing time. This read occurs somewhere between their addition to the queue in the AsyncPayloadProcessor but before `payloads.take()` is called. A hacky patch is to call `byteBuf = byteBuf.readerIndex(0);` as the first line in `RemotePayloadProcessor.encodeBinaryToString()`, to reset their reader index, and indeed this prevents the issue from arising.

### PR DESCRIPTION
**Pull request by njhill**
#### Motivation

An issue was reported in #111 where response payloads logged by a configured `PayloadProcessor` were sometimes incorrectly empty.

#### Modifications

Do not attempt to avoid slicing the response bytebuf in the case that a `PayloadProcessor` is configured, do not attempt to avoid some additional refcount updates in status != OK case.

#### Result

Hopefully no more incorrectly empty logged payloads

Resolves #111

### CODE REVIEW

**Comment by ckadner** [comment]
FYI @RobGeada

**Review by ckadner — APPROVED** [review]
@njhill -- your changes look good. I just rebased your branch to prepare for merging your PR, but I realized this PR should probably wait for a review/approval by @RobGeada or @tteofili ?

Should we create a ticket to track this?
> It would also be useful to have a unit test for this, but the tests included here don't exercise the actual bug. Ideally we'd have a test that actually runs a local modelmesh with an asyncprocessor configured and can trigger the problem (other unit tests already run modelmesh to test other stuff and could hopefully be used as a starting point and adapted).

I will remove my accidental co-authorship before merging this PR.

**Review by tteofili — APPROVED** [review]
LGTM

thanks @njhill and @ckadner

**Pull request merged**
